### PR TITLE
SI-4841 Plugins get a class path

### PR DESCRIPTION
**Pull request by som-snytt**
Let -Xplugin specify a class path (or multiple of them).

Each entry can be a jar or dir, and the first entry to supply
a plugin descriptor defines the plugin to load.

In the proposed scheme, each plugin gets a class loader so that
plugins are isolated from each other.

Presumably, if compiler plugins were a rich ecosystem, this would
have become a requirement by now.

Partest can't do multiple plugins yet, but this is what it looks like:

```
skalac -Xplugin:sample.jar:useful.jar:util,needful.jar:another.jar:util,needful.jar:util:exploded -Xplugin-require:sample,another,other foo.scala

skalac -Xplugin:sample.jar:useful.jar:util,needful.jar:another.jar:util,needful.jar:util:exploded -Xplugin-require:sample,other -Xplugin-disable:another foo.scala

skalac -Xplugin:sample.jar:useful.jar:util,sample.jar:useful.jar:util -Xplugin-require:sample foo.scala
```

The manual test shows three plugins with various permutations of jars and dirs.

The manual test demonstrates that plugins only see classes on their class path:

```
Initializing test.plugins.SamplePlugin
needful.Needful? Failure(java.lang.ClassNotFoundException: needful.Needful)
useful.Useful? Success(class useful.Useful)
Initializing more.plugins.AnotherPlugin
needful.Needful? Success(class needful.Needful)
useful.Useful? Failure(java.lang.ClassNotFoundException: useful.Useful)
Initializing other.plugins.OtherPlugin
```

Disabling a plugin results in a message instead of silent suppression:

```
Disabling plugin another
```

The duplicate plugin class test must still be honored:

```
Ignoring duplicate plugin sample (test.plugins.SamplePlugin)
Initializing test.plugins.SamplePlugin
needful.Needful? Failure(java.lang.ClassNotFoundException: needful.Needful)
useful.Useful? Success(class useful.Useful)
```

If the path is bad, then missing classes will report which plugin induced
the error:

```
Error: class not found: util/Probe required by test.plugins.SamplePlugin
Error: class not found: util/Probe required by more.plugins.AnotherPlugin
Initializing other.plugins.OtherPlugin
needful.Needful? Success(class needful.Needful)
useful.Useful? Failure(java.lang.ClassNotFoundException: useful.Useful)
error: Missing required plugin: sample
error: Missing required plugin: another
two errors found
```

fixes https://github.com/scala/bug/issues/4841

### CODE REVIEW

**Comment by som-snytt** [comment]
There is some cognitive dissonance in the plugin feature, in that a missing plugin is a warning unless you specify -Xplugin-require.  While the "extension directory" mechanism -Xpluginsdir does suggest I might be compiling with some unknown set of installed plugins (perhaps mandated by the team, for instance), if I ask for a plugin to run with -Xplugin, probably I really mean that I don't want to compile without it.

**Comment by som-snytt** [comment]
`typesafeDummy`?

**Comment by adriaanm** [comment]
our self-deprecating sense of humo(u)r -- part of the maven task that goes looking for an artifact. "Build failure" just means "artifact not found, kthx!"

**Comment by som-snytt** [comment]
PLS REBUILD/pr-scala@1c0c5ef

**Comment by som-snytt** [comment]
Rebased around the removal of `Pair`.

...and a partridge in a `Pair Tree`.

**Comment by som-snytt** [comment]
Updated not to bail on `-Xplugin:garbage`, which scala-ide tests for.  It will warn if `-Xdev`, because, dude, I said `-Xplugin` because I want a plugin.  There is no flag to turn on a message if `-Xplugindirs` is void of plugins.  Also added a test with two sample plugins going head-to-head.

File: src/compiler/scala/tools/nsc/plugins/Plugin.scala [hunk 1]
```diff
@@ -164,3 +189,11 @@ object Plugin {
     (clazz getConstructor classOf[Global] newInstance global).asInstanceOf[Plugin]
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/plugins/Plugin.scala:189]: Still throws on error instead of reporting.

**Comment by som-snytt** [comment]
Review by @adriaanm perhaps, who touched plugins code once has already commented in an aside.  Or, he is also good at delegating.

**Comment by som-snytt** [comment]
I see we were just bumped to critical.  Possibly, any issues under Si-5000 are grandfathered into criticality.  But Adriaan may be too busy delegating to delegate the review.  Plus I think I noticed he has an upcoming meetup.

**Comment by som-snytt** [comment]
I know @retronym enjoys a good test.  He has seen everything under the sun, no doubt; but this is more like strolling into a carnival to watch an amateur, not even a magician or high wire act, but some guy who bends over backward to walk on his hands.  That's kind of entertaining, right?

**Comment by adriaanm** [comment]
Criticality was my JIRA bulk edit device. Trying to machete my way through without turning into the hulk editor.

File: src/compiler/scala/tools/nsc/plugins/Plugin.scala [hunk 2]
```diff
@@ -113,11 +114,14 @@ object Plugin {
...
-      loader loadClass classname
-    } recoverWith {
-      case _: Exception =>
-        Failure(new RuntimeException(s"Warning: class not found: ${classname}"))
+    import scala.util.control.NonFatal
+    try {
+      Success[AnyClass](loader loadClass classname)
+    } catch {
+      case NonFatal(e) =>
+        Failure(new PluginLoadException(classname, s"Error: unable to load class: $classname"))
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/plugins/Plugin.scala:122]: `PluginLoadException` should wrap the underlying exception as its `cause`.

**Comment by retronym** [comment]
Highly entraining feats of contortionism. My only comment was really about old code, so:
LGTM

File: src/compiler/scala/tools/nsc/plugins/Plugin.scala (hunk 2 above)
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/plugins/Plugin.scala:122]: I was responsible for the previous iteration, too.  I don't remember before that (fail silently?) but I do remember wondering if people want to see a stack trace about a missing class, that says "ClassNotFound".  By people, I mean the guy who takes up support for continuations plugin and can't get the class path right in sbt, or whatever.  In Plugins, it recover[Any]s and just uses the message, which would include the causative message IIUC.

**Pull request merged**